enclave/crypto/gcm.c

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <mbedtls/cipher.h>
+
+#include <openenclave/internal/raise.h>
+#include <openenclave/internal/crypto/gcm.h>
+
+oe_result_t oe_aes_gcm_encrypt(
+    const uint8_t* key,
+    size_t key_size,
+    const uint8_t* iv,
+    size_t iv_size,
+    const uint8_t* aad,
+    size_t aad_size,
+    const uint8_t* input,
+    size_t inlen,
+    uint8_t* output,
+    uint8_t* tag)
+{
+    const mbedtls_cipher_info_t *info;
+    mbedtls_cipher_context_t gcm;
+    oe_result_t result = OE_OK;
+    size_t olen;
+
+    if (key_size != 16)
+        return OE_UNSUPPORTED;
+
+    mbedtls_cipher_init(&gcm);
+
+    info = mbedtls_cipher_info_from_type(MBEDTLS_CIPHER_AES_128_GCM);
+    if (info == NULL ||
+        mbedtls_cipher_setup(&gcm, info) ||
+        mbedtls_cipher_setkey(&gcm, key, (int)key_size * 8, MBEDTLS_ENCRYPT) ||
+        mbedtls_cipher_auth_encrypt(&gcm, iv, iv_size, aad, aad_size, input, inlen,
+                                    output, &olen, tag, 16))
```

Comment:
After reading up a bit, it looks 16 bytes is the smallest non-truncated tag. It may be useful to add a comment motivating this choice?

Reply:
For GCM, tag size should always be the same as block size. In the last commit, I replaced it with `info->block_size`.